### PR TITLE
Add an example esm-example package and use it in edit-post

### PR DESCRIPTION
**Pull request by youknowriad**
Related to #36716.
A simpler version of #34140 

The idea of this PR is simpler than the other one, instead of rewriting all the core modules to be esm modules and trying to be backward compatible, the following PR does this instead:

 - Adds the import map APIs to WP
 - Adds an example ESM-ready package that can be lazy loaded. (Called `esm-example) 
 - Lazy loads that package in `edit-post` (when loading the editor) as a proof of concept to ensure everything works as intended.

This is very promising and is almost ready IMO. Here are some of the tasks that I'd like to do to land it:

 - [x] Add a shim/polyfill to make it work for other browsers than Chrome. I have tried this one https://github.com/guybedford/es-module-shims which seems to be the most compelling one but failed to make it work entirely. I made the import map work but dynamic import `import( something )` no, I think we might need a web pack config to rewrite `import( something )` to `importShim( something )` but failed to do so. I could use some help to achieve this item.
 - [x] Right now, each ESM modules generates a dedicated entry in the import map, this work but I'd like to see if I can manage to have a single item for all. `@wordpress` esm packages. It might require generating files without the `.js` extension and putting these files in the same folder.
 - [ ] Right now the WP importmap is only loaded in the admin, I think we should load it in the frontend as well (to allow themes to use these modules and add their own modules)

### CODE REVIEW

File: packages/edit-post/src/index.js [hunk 1]
```diff
@@ -147,6 +147,9 @@ export function initializeEditor(
 	);
 }
 
+// esm-example
+import( '@wordpress/esm-example' ).then( ( mod ) => mod.run() );
```
Comment (gziolo) [inline packages/edit-post/src/index.js]: Is the following case in webpack plugin necessary for the import above?

```js
if ( ESM_PACKAGES.includes( request ) ) {
 	return 'import ' + request;
}
```

Comment (youknowriad) [reply packages/edit-post/src/index.js]: yes, this config ensures webpack keeps the import as is and do not try to bundle or resolve the `@wordpress/esm-example` dependency. 

Ideally, I'd like webpack to rewrite it `importShim( '@wordpress/esm-example' ).then( ( mod ) => mod.run() );` but I was not able to do so.

File: packages/esm-example/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# ESM Example
+
+This is an example of an ESM module. The module doesn nothing but it shows contributors how to build an esm ready package in Gutenberg.
```
Comment (sarayourfriend) [inline packages/esm-example/README.md]:
```suggestion
This is an example of an ESM module. The module does nothing but it shows contributors how to build an esm ready package in Gutenberg.
```

**Review by sarayourfriend — COMMENTED** [review]
I don't have much to add to this review, but it looks super interesting. What are the implications exactly?

**Comment by youknowriad** [comment]
@sarahmonster the implication is that we can start extracting code to lazy-loaded modules and have them work both in npm packages and in Gutenberg itself. There are several things we could explore later once this lands:

 - Explore extracting edit functions of large blocks (classic, ...) to separate lazy loaded packages
 - See if there are systematic ways to enable this for "edit" functions
 - Explore adding large lazy loaded dependencies (syntax highlighting is one that come to mind and that we delayed even before 5.0 because of that)
 - ...

**Comment by sarayourfriend** [comment]
That's awesome!

Just a note that the heaviest part of the classic block isn't anything implemented in Gutenberg, but the dependency on TinyMCE itself (as initialized by WP core). Has that changed? Otherwise there'd still be a lot of work necessary to async load the classic/freeform block.

**Comment by youknowriad** [comment]
> Has that changed? 

No, I actually didn't look at the classic block at all, I was thinking that if we find a way to make tinymce available as an esm package somehow we could lazy load it using this PR. That may be utopia, I don't know but it's just one use-case.

**Comment by sarayourfriend** [comment]
Yeah unfortunately it's a long shot... I spent a long time trying to figure out how to asynchronously load TinyMCE and it turns out there's a lot of work that WP core does to initialize TinyMCE (especially around i18n stuff) that become a blocker. Of course you could re-implement all that initialization client side.

It'd be a HUGE win to async load TinyMCE. That thing is a tremendous bloat for something that is so rarely used. I think it would have an actual measurable environmental impact considering the size and how often Gutenberg is loaded, not to mention a significant speed increase. More details about it here: https://github.com/WordPress/gutenberg/pull/29681

**Comment by youknowriad** [comment]
I have one good news and one bad news here:

 - The good news is that I was able to make the ship/polyfill work
 - The bad news is that I was not able to make it work while retaining the use of the default `import` function, instead I was forced to use `importShim` instead. This means we can't officially say that this is a public API for third-plugins to use because in the future, once the support is good enough, the ship will get removed. (maybe they can use at their own risk).

So basically, I think if folks like the approach, we can consider starting to use this in core code for some time.

**Comment by youknowriad** [comment]
> Right now, each ESM modules generates a dedicated entry in the import map, this work but I'd like to see if I can manage to have a single item for all. @wordpress esm packages. It might require generating files without the .js extension and putting these files in the same folder.

For this point, I read a bit and tried several approaches, I think extension less files might not be a great solution here. Let's start with one entry per module (like the standard recommends).

**Comment by tomalec** [comment]
> The bad news is that I was not able to make it work while retaining the use of the default import function, instead I was forced to use `importShim` instead. This means we can't officially say that this is a public API for third-plugins to use because in the future, once the support is good enough, the ship will get removed. (maybe they can use at their own risk).


I do agree this is a problem, but I think we can try somewhat sell it better. If we state to plugin developers something like:

"Hey! We are starting the process of modernizing our JS packages loading strategy, to use the latest and coolest ESM and import maps. Import-maps are not yet fully supported in all the browsers, so if you want to use the new ESM packages in your regular scripts, you would have to use `wpImport` instead of `import` so WP could apply the polyfill if needed."

- We will [rename `importShim`](https://github.com/guybedford/es-module-shims/issues/156#issuecomment-906570095) to something from WP namespace, so even once "once the support is good enough" and "the shim will get removed", we could keep aliasing the native `import`.

Now putting the plugin developer hat on, I'd read it as:

Oh, there is a breeze of new features coming. But for the old scripts, everything works as before, so I feel safe and calm.
Ah, and here is this new `@wordpress/syntax-highlighting` package, now shipped as a module. Ok, I'd import it slightly differently with `wpImport` to make sure it runs in my `<script>`-based plugin. Also, I'd start thinking of migrating my own code to ES modules, where I could import it natively.

-----

Forgive me if I missed something important here. But for me personally, it does not look that bad. The trade of using a new Web feature in the WordPress platform does not require any refactor on the plugin end but just to use slightly WordPress-specific syntax instead of native. That's only 2 characters difference, and WP will keep its compatibility for a long.

Plus, if I understood the OP correctly, this difference applies only to the new packages, that we will ship as ESM, not to the existing ones.

**Comment by youknowriad** [comment]
@tomalec your reasoning makes total sense to me yes, I just wanted to avoid that temporary step. But in the end it's not that bad, once it's well supported we could just do `wpImport = import` and call it a day.
